src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -149,6 +154,116 @@ public final EventLoop eventLoop() {
     }
 
     /**
+     * Returns a {@link EventExecutor} that will make sure this invocation is set
+     * as the current invocation before executing any callback. This should
+     * almost always be used for executing asynchronous callbacks in service
+     * code to make sure features that require the invocation context work
+     * properly. Most asynchronous libraries like
+     * {@link java.util.concurrent.CompletableFuture} provide methods that
+     * accept an {@link Executor} to run callbacks on.
+     */
+    public final EventExecutor contextAwareEventExecutor() {
+        return new ServiceInvocationContextAwareEventExecutor(this, eventLoop());
+    }
+
+    /**
+     * Returns a {@link Executor} that will execute callbacks in the given
+     * {@code executor}, making sure to propagate the current invocation context
+     * into the callback execution. It is generally preferred to use
+     * {@link #contextAwareEventExecutor()} to ensure the callback stays on the
+     * same thread as well.
+     */
+    public final Executor makeContextAware(Executor executor) {
+        return runnable -> executor.execute(makeContextAware(runnable));
+    }
+
+    /**
+     * Returns a {@link Callable} that makes sure the current invocation context
+     * is set and then invokes the input {@code callable}.
+     */
+    public final <T> Callable<T> makeContextAware(Callable<T> callable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                return callable.call();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link Runnable} that makes sure the current invocation context
+     * is set and then invokes the input {@code runnable}.
+     */
+    public final Runnable makeContextAware(Runnable runnable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                runnable.run();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link GenericFutureListener} that makes sure the current invocation
+     * context is set and then invokes the input {@code listener}.
+     */
+    public final <T extends Future<?>> GenericFutureListener<T> makeContextAware(
+            GenericFutureListener<T> listener) {
```

Comment:
Could we split this method to the one that accepts FutureListener and the one that accepts ChannelFutureListener? I know it will lead to code duplication, but GenericFutureListener is not intended to be used directly.